src/geometry/Lie_group/lie_group.lean

```diff
@@ -0,0 +1,224 @@
+
+* `Lie_add_group I G` : a Lie additive group where `G` is a manifold on the model with corners `I`.
+* `Lie_group I G`     : a Lie multiplicative group where `G` is a manifold on the model with
+                        corners `I`.
+
+## Implementation notes
+A priori, a Lie group here is a manifold with corner.
+-/
+
+section Lie_group
+
+universes u v
+
+/-- A Lie (additive) group is a group and a smooth manifold at the same time in which
+the addition and negation operations are smooth. -/
+class Lie_add_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (I : model_with_corners 𝕜 E E)
+  (G : Type*) [add_group G] [topological_space G] [topological_add_group G]
+  [charted_space E G] [smooth_manifold_with_corners I G] : Prop :=
+  (smooth_add : smooth (I.prod I) I (λ p : G×G, p.1 + p.2))
+  (smooth_neg : smooth I I (λ a:G, -a))
+
+/-- A Lie group is a group and a smooth manifold at the same time in which
+the multiplication and inverse operations are smooth. -/
+@[to_additive Lie_add_group]
+class Lie_group {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E]
+  [normed_space 𝕜 E] (I : model_with_corners 𝕜 E E)
+  (G : Type*) [group G] [topological_space G] [topological_group G]
+  [charted_space E G] [smooth_manifold_with_corners I G] : Prop :=
+  (smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2))
+  (smooth_inv : smooth I I (λ a:G, a⁻¹))
+
+section
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E] {I : model_with_corners 𝕜 E E}
+{F : Type*} [normed_group F] [normed_space 𝕜 F] {J : model_with_corners 𝕜 F F}
+{G : Type*} [topological_space G] [charted_space E G] [smooth_manifold_with_corners I G] [group G]
+[topological_group G] [Lie_group I G]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H' : Type*} [topological_space H'] {I' : model_with_corners 𝕜 E' H'}
+{M : Type*} [topological_space M] [charted_space H' M] [smooth_manifold_with_corners I' M]
+{E'' : Type*} [normed_group E''] [normed_space 𝕜 E'']
+{H'' : Type*} [topological_space H''] {I'' : model_with_corners 𝕜 E'' H''}
+{M' : Type*} [topological_space M'] [charted_space H'' M'] [smooth_manifold_with_corners I'' M']
+
+@[to_additive]
+lemma smooth_mul : smooth (I.prod I) I (λ p : G×G, p.1 * p.2) :=
+Lie_group.smooth_mul
+
+@[to_additive]
+lemma smooth.mul {f : M → G} {g : M → G} (hf : smooth I' I f) (hg : smooth I' I g) :
+  smooth I' I (f * g) :=
+smooth_mul.comp (hf.prod_mk hg)
+
+@[to_additive]
+lemma smooth_mul_left (a : G) : smooth I I (λ b : G, a * b) :=
+smooth_mul.comp (smooth_const.prod_mk smooth_id)
+
+/-- `L g` denotes left multiplication by `g` -/
+def L : G → G → G := λ g : G, λ x : G, g * x
+
+@[to_additive]
+lemma smooth_mul_right (a : G) : smooth I I (λ b : G, b * a) :=
+smooth_mul.comp (smooth_id.prod_mk smooth_const)
+
+/-- `R g` denotes right multiplication by `g` -/
+def R : G → G → G := λ g : G, λ x : G, x * g
```

Comment:
I want this to be valid only under the hypothesis of G being a Lie group, I do not care about unused arguments. How can I avoid Lint complaining? Also, should I protect it with a namespace?

Reply:
Why do you want this to be valid only under the hypothesis of G being a Lie group? And yes, it should be in a namespace.

Reply:
@Nicknamen I think that definition is also useful for any (non-Lie) group acting on itself. So please put it in a general spot. (And only assume `has_mul`.)
Once you've done that, maybe it should no longer be in a namespace (I wouldn't know which one) but give it a slightly longer name (like `right_mul`?)
Also, if you tag it with `@[to_additive]` then it also becomes available in the additive hierarchy.

Reply:
To disable the linter, write `@[nolint unused_arguments] def ...`. And yes, this should definitely be in a `Lie_group` namespace.

Reply:
I do not want to use it for general groups because indeed it is not common notations for general groups (as far as I saw), and I want to use exactly this notation for Lie groups because this is the most used notation in maths and it makes things way more readable. I do agree for general groups it'd be cool to have `right_mul` but my goal here is to make statements of theorem about left invariant vector fields and similar stuff readable as much as they are readable with this notation in common maths!

Reply:
You can have `left_mul` and `right_mul` as general concepts for general groups, and a localized notation that let's you write `L` and `R` in a Lie group context.

Reply:
Yeah sure! I prefer doing it only for Lie groups now because I do not really know the algebra folder and I do not know where to put such definitions but if someone really wants the general notion for any group with this PR please feel free to push a commit